### ISSUE TITLE
Filtering on fields of type dictionary does not work

### ISSUE THREAD

**Issue by torie** [rs/rest-layer#139]
I am trying to filter responses by a specific key-value combination for a dictionary field, which seems to not be working. 

## Steps to reproduce:

Define a resource with a schema.Field with the following specification:

```golang
"foobar": {
        Description: "A set of key values.",
        Validator: &schema.Dict{
	        KeysValidator:   &schema.String{MinLen: 1},
		ValuesValidator: &schema.String{MinLen: 1},
	},
	Filterable: true,
},
```

Make a request to get resources which has a specific key-value pair in the `foobar` dictionary.

```http
http://myapi.com/myresource?filter={"foobar":{"foo": "bar"}}
```

The following error occurs.

```
"Invalid `filter` parameter: foobar: invalid query expression: not a dict"
```

---

I've done some digging, and the error occurs in schema/dict.go when the `func (v Dict) Validate(...)`  function attempts to check if the incoming value is of type map[string]interface{}. At this point the value has been parsed into map[string]query.Value which fails the test.

```golang
// Validate implements FieldValidator interface.
func (v Dict) Validate(value interface{}) (interface{}, error) {
	dict, ok := value.(map[string]interface{}) // <- value is of type map[string]query.Value here
	if !ok {
		return nil, errors.New("not a dict")
	}
        ...
```

**Comment by smyrman** [comment]
For us (me and @torie at Searis), there are actually two cases we would like to work:

- Case 1:  exact match `http://myapi.com/myresource?filter={"foobar":{"foo": "bar"}}`

This is actually not too important.

- Case 2: search for the value of a key `http://myapi.com/myresource?filter={"foobar.foo": "bar"}`

This is more relevant, as it would allow us to use a dictionary to hold arbitrary tag/label values and filter based on individual fields.

**Comment by smyrman** [comment]
As for case2, the first step would be to make sure the following code can find the field of a `schemaDict` or `schema.Object`:

https://github.com/rs/rest-layer/blob/5b33327a0fb07b058fa7f55e5c7798c1a9b07164/schema/schema.go#L57

Most likely this would involve these two types implementing a `GetField` method and check for it in a similar way we already do if a field have a `Schema`.

**Comment by smyrman** [comment]
Starting on `case 2` only for now.

**Comment by smyrman** [comment]
> "Invalid `filter` parameter: foobar: invalid query expression: not a dict"

This issue is the same for both Dict and Array, and is caused by the parsing to `map[string]query.Value` and []quey.Value` over `map[string]interface{}` and `[]interface{}`.

https://github.com/rs/rest-layer/blob/14f963670c7da03bcd2989f16647a8944f04e8c7/schema/query/predicate.go#L68

According to @torie, the type does not seam to have any methods or other properties. Replacing it with `interface{}` in the code, would make the parsing simpler, solve ths bug, and deprecate functions that exist in the code today to translate between `query.Value` and `interface{}`. We could also use a go type alias over a declaration, if we want that for clarity.

```go
type Value = interface{}
```

**Comment by smyrman** [comment]
PS! This is for the exact match semantics.

**Comment by rs** [comment]
👌

**Comment by smyrman** [comment]
This should be completely solved by #177
